### PR TITLE
Storage FSM

### PR DESCRIPTION
**Pull request by artyom-yurin**
### Description of the Change
Storage FSM
All States and Events with action

### Benefits
Part of Sealing

### Possible Drawbacks 
None

### CODE REVIEW

File: core/miner/storage_fsm/impl/sealing_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  using primitives::address::Address;
+  using primitives::piece::PaddedPieceSize;
+  using primitives::piece::PieceInfo;
+  using primitives::piece::UnpaddedPieceSize;
+  using proofs::SealRandomness;
+  using sector_storage::InteractiveRandomness;
+  using sector_storage::Manager;
+  using sector_storage::PreCommit1Output;
+
+  struct SectorInfo {
```
Comment (Alexey-N-Chernyshov) [inline core/miner/storage_fsm/impl/sealing_impl.hpp:34]: Is it the same is core/primitives/sector/sector.hpp:151?

File: core/miner/storage_fsm/api.hpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+  template <typename... T>
+  using ParamsTuple =
+      std::tuple<std::remove_const_t<std::remove_reference_t<T>>...>;
+
+  class SealingApi {
+   public:
+    struct ChainHeadResponse {
+      TipsetToken tipset;
+      ChainEpoch epoch;
```
Comment (Alexey-N-Chernyshov) [inline core/miner/storage_fsm/api.hpp:49]: height is stored in tipset

File: core/miner/storage_fsm/impl/sealing_impl.hpp (hunk 1 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/impl/sealing_impl.hpp:34]: No, this is context object between FSM states

File: core/miner/storage_fsm/api.hpp (hunk 2 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/api.hpp:49]: Yes, but here tipset is just key of real tipset. I suppose to choose better name or return Full tipset

File: core/miner/storage_fsm/api.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+               const Address &,
+               const TipsetToken &);
+
+    API_METHOD(StateMinerInitialPledgeCollateral,
+               TokenAmount,
+               const Address &,
+               SectorNumber,
+               const TipsetToken &);
+
+    struct MessageReceipt {
```
[the remark this answers is not in the document]
Comment (artyom-yurin) [reply core/miner/storage_fsm/api.hpp:84]: Yes, it is duplicate. I will remove my type

File: core/miner/storage_fsm/api.hpp (hunk 2 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/api.hpp]: Removed

File: core/miner/storage_fsm/api.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (artyom-yurin) [reply core/miner/storage_fsm/api.hpp]: Removed

**Pull request merged**